### PR TITLE
Fix starting order for `RollbackSynchronizer`

### PR DESCRIPTION
**Pull request by TheYellowArchitect**
Solves #260 and #268 by fixing input initialization.

Also gets rid of the `await 1 frame` hack for 2 seperate classes (BrawlerController.gd/player.gd) in order to await for input authority to be set before `process_authority`, so this awaiting a frame logic is moved to `RollbackSynchronizer` since in most games, input authority is set in `_ready` or something similar. 

May also solve the reported problem of the first 2 ticks, having NPCs spinning. Haven't seen it in a project of mine, but it may solve it because this PR addresses the first `_earliest_input` being set properly

### CODE REVIEW

File: addons/netfox/rollback/rollback-synchronizer.gd [hunk 1]
```diff
@@ -24,6 +26,7 @@ var _states: Dictionary = {}
 var _inputs: Dictionary = {}
 var _latest_state: int = -1
 var _earliest_input: int
+var _started_at_tick: int = -1
```
Comment (TheYellowArchitect) [inline addons/netfox/rollback/rollback-synchronizer.gd:29]: This variable ultimately ended up being unused, but its useful to keep so as to know when an avatar joined/started.

File: addons/netfox/rollback/rollback-synchronizer.gd [hunk 2]
```diff
@@ -84,13 +83,24 @@ func process_authority():
...
+	_earliest_input = NetworkTime.tick
+	if (_started_at_tick == -1):
+		_started_at_tick = NetworkTime.tick
+
 func _ready():
 	process_settings()
+
+	if (await_first_frame):
+		await get_tree().process_frame
+	process_authority()
```
Comment (TheYellowArchitect) [inline addons/netfox/rollback/rollback-synchronizer.gd:97]: After `await`, it is basically guaranteed the input authority is set, so `process_authority` can be safely invoked

File: addons/netfox/rollback/rollback-synchronizer.gd [hunk 3]
```diff
@@ -220,6 +232,8 @@ func _get_history(buffer: Dictionary, tick: int) -> Dictionary:
 	var latest = buffer.keys().max()
 
 	if tick < earliest:
+		_logger.warning("Tried to load tick %s which is earlier than the earliest we have recorded (%s)
+		Try increasing the history limit." % [tick, earliest])
```
Comment (TheYellowArchitect) [inline addons/netfox/rollback/rollback-synchronizer.gd:236]: Debug comment, to confirm #260  is fixed. Place this message on main branch, and you will see it goes from 0 to current...

File: examples/forest-brawl/scripts/brawler-controller.gd [hunk 4]
```diff
@@ -47,10 +47,6 @@ func _ready():
 	
 	_snap_to_spawn()
 
-	# TODO: What if the RollbackSynchronizer had a flag for this?
-	# Wait a frame so Input has time to get its authority set
```
Comment (TheYellowArchitect) [inline examples/forest-brawl/scripts/brawler-controller.gd:51]: And that's what I did :P

**Comment by TheYellowArchitect** [comment]
Btw the "await 1 frame" logic could be skipped entirely if `RollbackSynchronizer` is below (or above?) the Input node(s), so Input's `_ready` fires before `RollbackSynchronizer`'s. I don't remember whether Godot's tree load order, is down to up or up to down, but assuming up to down, then `RollbackSynchronizer` being below input node(s) or at the bottom, the same result as now occurs, without those 3 lines (and an ugly exported boolean) - but it must be communicated to the user via docs or comments.

File: examples/forest-brawl/scripts/brawler-controller.gd (hunk 4 above)
Comment (elementbound) [reply examples/forest-brawl/scripts/brawler-controller.gd:51]: Turns out this is not needed at all, we can just defer the `process_settings()` call in RollbackSynchronizer - running it in the next idle time, usually at the end of the current `_process()`.

**Comment by elementbound** [comment]
Updated PR, turns out the single-frame wait is not needed at all. Also explained it under https://github.com/foxssake/netfox/issues/260#issuecomment-2464755741

**Comment by TheYellowArchitect** [comment]
There is a weird "bug" which I have to mention. For the first second, sometimes, the client hovers. Here is a video example of this: https://youtu.be/BZGODoPt_6w
I tried removing the deferred call and straight up adding an await, but the result was the same.
It is not an issue blocker, because without this PR there is straight up desync lol

To confirm this bug isn't a regression, I checked out to commit https://github.com/foxssake/netfox/pull/284/commits/20c5bfd97d91e6f279acabcc4000797060636db4
It was very rare compared to this commit, but I did manage to reproduce it at that commit too.

I assume it has something to do with `is_initialized` and should be `visible = false` until then? idk, but I like the idea of `is_initialized` you added.

**Comment by TheYellowArchitect** [comment]
On the addition of `is_initialized`, since its a boolean, might as well convert it to integer to track which tick the rollback synchronizer started. The default value being -1. So the condition comparison is `if started_at > -1`

**Pull request merged**